keps/sig-windows/4802-windows-node-shutdown/README.md

```diff
@@ -0,0 +1,879 @@
+
+While for windows service programs, we will have more freedom to register a service control handler to monitoring various shutdown related event 
+such as "SERVICE_CONTROL_PRESHUTDOWN", "SERVICE_CONTROL_SHUTDOWN". So in order to configure and process the (pre)shutdown event from windows system, 
+register the kubelet application as a windows service program will be the most straightforward way.
+
+Aso from the same document [SetConsoleCtrlHandler function](https://learn.microsoft.com/en-us/windows/console/setconsolectrlhandler?redirectedfrom=MSDN),
+It appears that "SERVICE_CONTROL_PRESHUTDOWN" is more flexible than "SERVICE_CONTROL_SHUTDOWN" for a windows service to monitor and control
+
+### Implementation
+Similar with the linux case, "nodeshutdown_manager_windows.go" will be added to replace the current
+dumb implementation for windows, it will act as the main role in the node shutdown scenario and go through similar 
+initialization/start process during kublet start up.
+
+Also, we will update the windows service package to "AcceptPreShutdown" and add a pre-shutdown handler to process
+the "SERVICE_CONTROL_PRESHUTDOWN" event. This pre-shutdown handler will have an interface "ProcessShutdownEvent",
+which will be implemented by the "nodeshutdown_manager_windows"
+
+The entry point of running the kublet as a windows service is "InitService", in which, we can initialize and register 
+the servcie control handler, the above "pre-shutdown handler" is part of this service control handler.
+
+At this "InitService" stage, the kubelet has not read the kubelet configs/flags yet, so the "nodeshutdown_manager_windows", 
+has not been instantiated, we will then set the "preshutdownhander" to be NIL, so will ignore the preshutdown event if any. 
+This should be safe enough as the kubelset itself has not been started, let alone the pods it managed, so there is nothing 
+need to do for shutting down at this stage.
+
+When kubelet proceeds with starting up and initialize the "nodeshutdown_manager_window", it will set the servic control ehandler’s 
+pre-shutdownhandler to the "nodeshutdown_manager_window" and is now ready to monitor and process the preshutdown event
+
+We will also add function of "UpdatePreShutdownInfo" and "QueryPreShutdownInfo" to windows service package, these functions will be 
+used by the "nodeshutdown_manager" to query and update the preshtudown timeout.
+
+If we will take the 'pre-shutdown timeout value as pariale with the 'InhibitDelayMaxSec` in linux,
+then the other parts of implementation will be pretty much similar with what linux has [implemented](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2000-graceful-node-shutdown#implementation)
+
+The poc of the implementation can be found [here](https://github.com/zylxjtu/kubernetes/commit/854ea4bde88c0905241b43f5f80d470967bb909f#diff-8494875e9a1e884afd36f32fa90dceb2b827616f1e981ecb83a12901262214c7)
+
+### Notes/Constraints/Caveats (Optional)
```

Comment:
Please add a note saying that we are aware alternate service managers on Windows like nssm.exe handle things like log-rotate for us and updating the logging mechanisms to handles this is our of scope for this kep but we'll look into other tools like the kube-runner as alternatives.

Reply:
will add